include/libebpf.h

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: LicenseRef-Elastic-License-2.0
+
+/*
+ * Copyright 2021 Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License 2.0;
+ * you may not use this file except in compliance with the Elastic License 2.0.
+ */
+
+// This file is just a spec for discussion, not included in builds or anything.
+// This does not compile and it's probably not in the folder we want it in.
+
+#ifndef LIBEBPF_H_
+#define LIBEBPF_H_
+
+#include <stdint.h>
+#include <sys/types.h>
+
+/* Kernel Features */
+enum ebpf_kernel_features {
+    EBPF_KERNEL_CAPABLE_BPF = (1 << 0),
+    EBPF_KERNEL_CAPABLE_RINGBUF = (1 << 1),
+    EBPF_KERNEL_CAPABLE_BTF = (1 << 2),
+};
+
+int ebpf_kernel__detect_features();
+
+/* Events fetching and manipulation */
+typedef enum ebpf_event_type {
+    EBPF_EVENT_TYPE_UNSPEC = 0,
+    EBPF_EVENT_TYPE_NETWORK_UPDATE = 1,
+    EVENT_NETWORK_IPV4_CONNECTION_ATTEMPTED = 2,
+    EVENT_NETWORK_IPV4_DISCONNECT_RECEIVED  = 3,
+    EVENT_NETWORK_IPV4_CONNECTION_ACCEPTED  = 4,
+    EVENT_NETWORK_IPV4_RECONNECT_ATTEMPTED  = 5,
+    EVENT_NETWORK_IPV4_HTTP_REQUEST         = 6,
+    EVENT_NETWORK_IPV6_CONNECTION_ATTEMPTED = 7,
+    EVENT_NETWORK_IPV6_DISCONNECT_RECEIVED  = 8,
+    EVENT_NETWORK_IPV6_CONNECTION_ACCEPTED  = 9,
+    EVENT_NETWORK_IPV6_RECONNECT_ATTEMPTED  = 10,
+    EVENT_NETWORK_IPV6_HTTP_REQUEST         = 11,
+    EVENT_PROCESS_FORK,
+    EVENT_PROCESS_EXEC,
+    EVENT_PROCESS_EXIT,
+    EVENT_PROCESS_TTY_OUTPUT,
+    EBPF_EVENT_MAX
+} ebpf_event_type;
+
+struct ebpf_event {
+    uint64_t ts;
+    ebpf_event_type type;
+    void *data;
+};
+
+/* opaque context, will contain union with struct perf_buffer,
+ * struct ring_buffer,
+ *
+ * we need a way to map from capability/event sets to probe sets,
+ * and we could store the cap/event sets directly or cache the result
+ * of the mapping in the context */
+typedef struct ebpf_event_ctx;
+
+typedef int (*event_handler_fn)(struct ebpf_event *event, size_t);
+
+typedef uint64_t ebpf_cap_set;
+
+/* result will be ((ACTUAL AND INTERESTED) OR FORCED) */
+ebpf_cap_set ebpf_event__detect_capabilities(
+        ebpf_cap_set interested,
+        ebpf_cap_set forced)
+
+/* Combination of events user is interested in. Mapping of events
+ * to probe that provides that event is stored internally. User
+ * does not need to know which probes to load */
+typedef uint64_t ebpf_event_set;
```

Comment:
Is this intended to be a bitmask? 

`ebpf_event_type` is not in powers of two. It would also be impractical in this case since we really want a list of event types and not something to compare against.

You probably wanted something like

```suggestion
typedef uint64_t ebpf_event_set[EBPF_EVENT_MAX];
```

Reply:
Perhaps the event types *should* be powers of two. We want to pass many of them at the same time, they are effectively a bitmask.

I like a bitmask because the arguments in other functions are then `uint64_t`s. However, that limits us to 64 events types without having to change the API. Perhaps that's enough for a long time.

Alternatively, as you suggest, we could use an array of `uint8_t`s with  `EBPF_EVENT_MAX` set to something like 1024. That would provide more runway, but then the APIs would have pointers for parameters. Array parameters (pointers) are easy for library users to get wrong and hard for library authors to validate.

Reply:
The advantages of having a bitmask are:

- efficient bitwise comparisons
- single value to represent a container of many states
- Very low footprint memory
 
Disadvantages

- Limited to 64 elements


Advantages of an array
- Can represent more than 64 elements

Disadvantages of an array
- More difficult to use in the API for the final user (they have to fill an array)


I honestly would go for a bitmask as well, question is: 64 elements are enough? And if it is, how much room do we have for new events?

Reply:
I'm pushing a change to fix the events to be a power of two since this is the way most people seem to like. We will see if 64 elements is enough while working on it.